tools/pythonpkg/src/python_import_cache.cpp

```diff
@@ -36,6 +36,7 @@ PyObject *PythonImportCacheItem::AddCache(PythonImportCache &cache, py::object o
 void PythonImportCacheItem::LoadModule(const string &name, PythonImportCache &cache) {
 	load_attempted = true;
 	try {
+		py::gil_assert();
```

Comment:
If I understand this correctly, `gil_check()` returns if the thread is holding the GIL or not, right?
Wouldn't it be the case that if it's not we just acquire the GIL for the Load Module function?
As in replace this with an if ...

Reply:
Yea I also thought of this, but Mark correctly pointed out that would cause us to silently degrade our GIL handling code, creating a lot of locks and unlocks instead of just one of each

Reply:
But maybe in the case of LoadModule it might be fine?

Reply:
I would say we should just never call `LoadModule` outside of places where we hold the GIL. This is likely to hide other bugs as well. For example, we might do something like `arrow().attr("function")()`. Now the `arrow` load correctly grabs the GIL but the function call does not - hiding a bug.

I would say perhaps this should not only be an assertion but also be checked in release mode/throw an internal exception. I don't think our Python code runs with assertions in the CI, at least not for all tests.

Reply:
Fair enough, I think internal exception sounds excellent.

Reply:
@Mytherin 
```c++
	if (!gil_check()) {
		throw duckdb::InternalException("The GIL should be held for this operation, but it's not!");
	}
```